File: 0013-roman-to-integer/0013-roman-to-integer.cpp
class Solution {
public:
    int romanToInt(string s) {
        unordered_map<char, int> mp;
        mp['I'] = 1;
        mp['V'] = 5;
        mp['X'] = 10;
        mp['L'] = 50;
        mp['C'] = 100;
        mp['D'] = 500;
        mp['M'] = 1000;

        int n = s.length();
        long long num = 0;
        int i = 0;

        // while(i < n)
        // {
        //     if(i + 1 == n)
        //     {
        //         num += (mp[s[i]]);
        //         i++;
                
        //     }

        //     else if(mp[s[i]] >= mp[s[i+1]])
        //     {
                
        //         num += (mp[s[i]]);
        //         i++;
        //     }

        //     else if(mp[s[i]] < mp[s[i+1]])
        //     {
        //         num += ( (mp[s[i+1]]) - (mp[s[i]]) );
        //         i = i + 2;

        //     }
        // }

        // return num;

        unordered_map<char, int> m;
        
        m['I'] = 1;
        m['V'] = 5;
        m['X'] = 10;
        m['L'] = 50;
        m['C'] = 100;
        m['D'] = 500;
        m['M'] = 1000;
        
        int ans = 0;
        
        for(int i = 0; i < s.length(); i++){
            if(m[s[i]] < m[s[i+1]]){
                ans -= m[s[i]];
            }
            else{
                ans += m[s[i]];
            }
        }
        return ans;
    }
    
};

File: 0013-roman-to-integer/0013-roman-to-integer.java
class Solution {
    public int romanToInt(String s) {
        HashMap<Character, Integer> map = new HashMap<Character, Integer>();

        map.put('I',1);
        map.put('V',5);
        map.put('X',10);
        map.put('L',50);
        map.put('C',100);
        map.put('D',500);
        map.put('M',1000);

        int sum = 0;
        int n = s.length();

        String str = "IV IX XL XC CD CM";

        for(int i = n-1; i >= 0;)
        {
            if(i - 1  >= 0 )
            {
                char curr = s.charAt(i);
                char prev = s.charAt(i-1);
            StringBuilder sb = new StringBuilder("");

                sb.append(prev);
                sb.append(curr);

                if(str.contains(sb.toString()))
                {
                    int val = (map.get(curr)) - (map.get(prev));
                    sum += val;
                    i = i - 2;
                }
                else
                {
                    sum += (map.get(curr));
                    i--;

                }
            }
            else
            {
                char curr = s.charAt(i);
                int val = map.get(curr);
                sum += val;
                i--;
            }

            System.out.println("i = " + i + " at sum = " + sum );
        }

        return sum;
    }
}